src/network/tarcap/packets_handler/handlers/dag_packets_handler.cpp

```diff
@@ -0,0 +1,230 @@
+                                     std::shared_ptr<SyncingState> syncing_state,
+                                     std::shared_ptr<TransactionManager> trx_mgr,
+                                     std::shared_ptr<DagBlockManager> dag_blk_mgr, std::shared_ptr<DbStorage> db,
+                                     uint16_t network_min_dag_block_broadcast, uint16_t network_max_dag_block_broadcast,
+                                     const addr_t &node_addr)
+    : PacketHandler(std::move(peers_state), node_addr, "DAG_BLOCK_PH"),
+      syncing_state_(syncing_state),
+      trx_mgr_(trx_mgr),
+      dag_blk_mgr_(dag_blk_mgr),
+      db_(db),
+      urng_(std::mt19937_64(std::random_device()())),
+      network_min_dag_block_broadcast_(network_min_dag_block_broadcast),
+      network_max_dag_block_broadcast_(network_max_dag_block_broadcast) {}
+
+void DagPacketsHandler::process(const PacketData &packet_data, const dev::RLP &packet_rlp) {
+  if (packet_data.type_ == NewBlockPacket) {
+    processNewBlockPacket(packet_data, packet_rlp);
+  } else if (packet_data.type_ == NewBlockHashPacket) {
+    processNewBlockHashPacket(packet_data, packet_rlp);
+  } else if (packet_data.type_ == GetNewBlockPacket) {
+    processGetNewBlockPacket(packet_data, packet_rlp);
+  } else {
+    assert(false);
+  }
+}
+
+inline void DagPacketsHandler::processNewBlockPacket(const PacketData &packet_data, const dev::RLP &packet_rlp) {
+  // Ignore new block packets when syncing
+  if (syncing_state_->is_syncing()) return;
+
+  DagBlock block(packet_rlp[0].data().toBytes());
+  blk_hash_t const hash = block.getHash();
+
+  if (dag_blk_mgr_) {
+    if (dag_blk_mgr_->isBlockKnown(hash)) {
+      LOG(log_dg_) << "Received NewBlock " << hash.toString() << "that is already known";
+      return;
+    }
+    if (auto status = syncing_state_->checkDagBlockValidation(block); !status.first) {
+      LOG(log_wr_) << "Received NewBlock " << hash.toString() << " missing pivot or/and tips";
+      status.second.push_back(hash);
+      syncing_state_->requestBlocks(packet_data.from_node_id_, status.second,
+                                    GetBlocksPacketRequestType::MissingHashes);
+      return;
+    }
+  }
+
+  const auto transactions_count = packet_rlp.itemCount() - 1;
+  LOG(log_dg_) << "Received NewBlockPacket " << transactions_count;
+
+  std::vector<Transaction> new_transactions;
+  for (size_t i_transaction = 1; i_transaction < transactions_count + 1; i_transaction++) {
+    Transaction transaction(packet_rlp[i_transaction].data().toBytes());
+    new_transactions.push_back(transaction);
+    peer_->markTransactionAsKnown(transaction.getHash());
+  }
+
+  peer_->markBlockAsKnown(hash);
+  if (block.getLevel() > peer_->dag_level_) peer_->dag_level_ = block.getLevel();
+  onNewBlockReceived(block, new_transactions);
+}
+
+inline void DagPacketsHandler::processNewBlockHashPacket(const PacketData &packet_data, const dev::RLP &packet_rlp) {
```

Comment:
I think that in general we could get rid of whole "NewBlockHashPacket" packet. When I checked it is only send inside "onNewBlockVerified" to "in some way randomized" set of peers that do not have this block already(or at least we did not mark it as known fro them) and I am not sure what is the purpose of it ? If such peer received only block hash but does not have it locally it will instantly request our node to send him full block. I guess it has some purpose but I am not really sure what it is ??? @justinsnapp @mfrankovi you might know ?

Reply:
comment moved to https://github.com/Taraxa-project/taraxa-node/issues/1038